src/vcu/src/controls.c

```diff
@@ -0,0 +1,191 @@
+          lc_state = SPEED_CONTROLLER;
+          printf("[LAUNCH CONTROL] SPEED CONTROLLER STATE ENTERED\r\n");
+        }
+        break;
+      case SPEED_CONTROLLER:
+        speed_command = get_launch_control_speed(front_wheel_speed);
+        sendSpeedCmdMsg(speed_command, torque_command);
+
+        // Transition
+        if (any_lc_faults()) {
+          lc_state = ZERO_TORQUE;
+          printf("[LAUNCH CONTROL] ZERO TORQUE STATE ENTERED\r\n");
+        }
+      case ZERO_TORQUE:
+        sendTorqueCmdMsg(0);
+
+        if (pedalbox_max(accel) < LC_ACCEL_RELEASE) {
+          lc_state = DONE;
+          printf("[LAUNCH CONTROL] DONE STATE ENTERED\r\n");
+        }
+        break;
+    }
+
+  } else {
+    // Control regen brake valve:
+    bool brake_valve_state = control_settings.using_regen && get_pascals(pedalbox.REAR_BRAKE) < RG_REAR_BRAKE_THRESH;
+    set_brake_valve(brake_valve_state);
+
+    int32_t regen_torque;
+    if (brake_valve_state) {
+      regen_torque = get_regen_torque();
+    } else {
+      regen_torque = 0;
+    }
+
+    // Extra check to ensure we are only sending regen torque when allowed
+    if (torque_command == 0) {
+      torque_command = regen_torque;
+    }
+
+    sendTorqueCmdMsg(torque_command);
+  }
+}
+
+static int16_t get_torque(void) {
+  auto accel = pedalbox_avg(accel);
+
+  if (accel < PEDALBOX_ACCEL_RELEASE) return 0;
+
+  int16_t raw_torque =  MAX_TORQUE * (accel - PEDALBOX_ACCEL_RELEASE) / (MAX_ACCEL_VAL - PEDALBOX_ACCEL_RELEASE);
+  return raw_torque * control_settings.limp_factor / 100;
+}
+
+static int32_t get_regen_torque() {
+  int32_t regen_torque = 0;
+
+  // IMU velocity component unit conversion:
+  // Starting units: 100 * m/s, ending units: kph
+  // (100 m/s) * (1 km/1000 m) * (3600 s/1 hr) = 360 kph
+  // So multiply by 360 to get kph --> but this is squared, so multiply by 129600
+  int32_t imu_vel_north_squared = imu_velocity.north * imu_velocity.north;
+  int32_t imu_vel_east_squared = imu_velocity.east * imu_velocity.east;
+  int32_t car_speed_squared = 129600 * (imu_vel_north_squared + imu_vel_east_squared);
+
+  if ((mc_readings.speed * -1 > RG_MOTOR_SPEED_THRESH) &&
+      (cs_readings.V_bus < RG_BATTERY_VOLTAGE_MAX_THRESH) &&
+      (get_pascals(pedalbox.FRONT_BRAKE) > RG_FRONT_BRAKE_THRESH) &&
+      car_speed_squared > 25) { // Threshold is 5 kph, so 25 (kph)^2
+    int32_t kilo_pascals = get_pascals(pedalbox.FRONT_BRAKE) / 1000;
+    // Because we already divded by 1000 by using kilo_pascals instead of
+    // pascals, we only need to divde by 10^4, not 10^7 for RG_10_7_K
+    // And by dividing by 10^3 instead of 10^4, we are multiplying by 10 to get
+    // dNm instead of Nm
+    regen_torque = RG_10_7_K * kilo_pascals * (100 - control_settings.cBB_ef) /(control_settings.cBB_ef * 1000);
+    if (regen_torque > RG_TORQUE_COMMAND_MAX) {
+      regen_torque = RG_TORQUE_COMMAND_MAX;
+    }
+  }
+
+  return -1 * regen_torque;
```

Comment:
Why is this negative?

Reply:
Because regen is negative torque
I suppose I can add a comment

Reply:
oh pls thx